### ISSUE TITLE
Extended Key Usage Cert with anyExtendedKeyUsage failing to connect

### ISSUE THREAD

**Issue by kulpaj** [openssl/openssl#17959]
Dear all, 
I've noticed it in openssl.1.1.1l but it seems that the same code exists also in most recent master repository

The issue I've encountered is having rejected connection due to X509_V_ERR_INVALID_PURPOSE on end entity certificate. (backtrace starts with SSL_accept(...) ) 

It happens when the only extended key usage set in the certificate is "any" 
If I understand correctly (https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.12) id-kp-clientAuth shall not be necessary on the client certificate if anyExtendedKeyUsage is used. 

It boils down to v3_purp.c file and call
if (xku_reject(x, XKU_SSL_CLIENT)) 
within 
static int check_purpose_ssl_client(...)
not checking for XKU_ANYEKU. 

Is it an intended behaviour, or an oversight?

**Comment by kaduk** [comment]
> If I understand correctly (https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.12) id-kp-clientAuth shall not be necessary on the client certificate if anyExtendedKeyUsage is used.

I think that's probably an oversimplification of the anyExtendedKeyUsage semantics.  Quoting from the linked resource:

```
                             Applications that require the presence of a
   particular purpose MAY reject certificates that include the
   anyExtendedKeyUsage OID but not the particular OID expected for the
   application.
```
Because openssl is a general-purpose library and there are situations in which accepting anyEKU is risky, I believe that the default behavior is consciously to not accept anyExtendedKeyUsage as a stand-in for any specific extended key usage.  This is evident from the checks performed by the "standard" (built-in) `X509_PURPOSE` methods defined in v3_purp.c -- they all explicitly look for the specific EKU defined for that purpose, and only the explicit `X509_PURPOSE_ANY` purpose will allow the "anyExtendedKeyUsage" to be considered valid.

**Comment by kulpaj** [comment]
While I understand secure by default approach, I failed to find a way to out-out from this measure. At least while using SSL_accept() to accept clients with no XKU_SSL_CLIENT extension set or to override purposed check.

On the other hand substituting it with X509_PURPOSE_ANY may set the security too loose (will setting conn->ssl->purpose to 7 prior to calling SSL_accept(conn) would trigger the X509_PURPOSE_ANY check?), as you would still like to check ku for KU_DIGITAL_SIGNATURE and KU_KEY_AGREEMENT.  

Is the only way to opt-out in this scenario to use X509_PURPOSE_add() call?

**Comment by kaduk** [comment]
You can also override the default verification callback.
If I remember correctly (which is questionable, as the API surface here is complicated and confusing), this is better done via `SSL_CTX_set_cert_verify_callback()` than `SSL_CTX_set_verify()`, the disclaimer in the former notwithstanding.
There was also a bit of previous discussion that's on a related/adjacent topic at https://github.com/openssl/openssl/issues/13965#issuecomment-776249427 linking to https://mailarchive.ietf.org/arch/msg/dtn/G2qi9Qwz09kkuzRbdEWrDqoHLwc/

**Comment by kulpaj** [comment]
SSL_CTX_set_cert_verify_callback is way up the stack and it seems not worth manipulating (too high risk and later maintenance cost to keep up to date, and although I am talking about EE certificate, the whole 3-level CA chain needs to be verified). 

I think the most reasonable solution for my problem is to add check_purpose_relaxed_ssl_client that allows sole "Any" xku and adding callback via X509_PURPOSE_add(), as that adds minimal and easily-understandable code. 

Is there a good doc on what name, sname, and arg are for X509_PURPOSE_add() and how to force SSL_accept to designate given purpose for EE certificate?

**Comment by kaduk** [comment]
> Is there a good doc on what name, sname, and arg are for X509_PURPOSE_add() and how to force SSL_accept to designate given purpose for EE certificate?

I don't know of one.
If you do go down this route, can I suggest that you start by writing such documentation and submitting them as a PR to this repo (i.e., POD files in doc/man3/)?  Then we would be able to review them and confirm that your understanding is correct before you start using the code.

**Comment by vdukhovni** [comment]
The best way to allow *any* EKU, is to not include the EKU extension. Then there are no constraints imposed.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input
